design/ratelimit-design-new.md

```diff
@@ -0,0 +1,180 @@
+
+## Detailed Design
+
+### RateLimitPolicy type
+
+A sample `RateLimitPolicy` looks like:
+
+```yaml
+rateLimitPolicy:
+  # local defines local rate limiting properties for the virtual host or route.
+  local:
+    # tokenBucket defines the properties of the local rate limiting token bucket.
+    # ref. https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/token_bucket.proto#envoy-v3-api-msg-type-v3-tokenbucket
+    tokenBucket:
+      maxTokens: 100
+      tokensPerFill: 10
+      fillInterval: 1s
+  # global defines global rate limiting properties for the virtual host or route.
+  global:
+    # rateLimitDescriptors defines the descriptors to be generated and sent to the
+    # external Rate Limiting Service (RLS) for a rate limit decision.
+    rateLimitDescriptors:
+    # This descriptor is generated only if the x-steve-ratelimit header
+    # is present on the request.
+    - items:
+        # adds a descriptor entry with a key of "generic_key" and a value
+        # of "s1".
+        - genericKey:
+            value: s1
+
+        # adds a descriptor entry with a key of "remote_address" and a value
+        # equal to the trusted address from x-forwarded-for.
+        - remoteAddress: {}
+
+        # adds a descriptor entry with a key of "steve-ratelimit"
+        # and a value equal to the value of the "x-steve-ratelimit" header.
+        - requestHeader:
+            descriptorKey: steve-ratelimit
+            headerName: x-steve-ratelimit
+        
+    # This descriptor is always generated since it's just a static value.
+    - items:
+        - genericKey:
+            value: s1
+```
+
+This API maps closely to Envoy's rate limiting API.
+
+For local rate limiting, the rate limit itself is defined inline in the `RateLimitPolicy`, as token bucket properties.
+Since local rate limits apply *per Envoy*, each pod in the Envoy daemon set will have a token bucket with these properties.
+
+For global rate limiting, per Envoy's API, the `RateLimitPolicy` only defines a list of descriptors to send to an external RLS for a given request.
+All [descriptor entries defined by Envoy](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-ratelimit-action) will be supported except `metadata` and `dynamic_metadata`.
+
+### Contour Configuration
+If using global rate limiting, the details of the external RLS must be configured with Contour.
+First, an `ExtensionService` must be defined for the RLS with the cluster-level details for the RLS itself.
+For example:
+```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: ExtensionService
+metadata:
+  namespace: projectcontour
+  name: ratelimit
+spec:
+  protocol: h2
+  services:
+    - name: ratelimit
+      port: 8081
+  timeoutPolicy:
+    response: 50ms
+```
+
+Second, a new block will be added to the Contour config file to identify the RLS extension service and provide some additional parameters:
```

Comment:
1. This assumes a single cluster will only have one RLS? (Maybe good to call out in the goals). I'm thinking possibly one team may want their own implementation of a global RLS different than another. 
2. It's unfortunate to need another config place. It seems like the ExtensionService should be where you define the domain, denyOnFailure, etc. Having an `RateLimitExtensionService` and a `AuthorizationExtensionService` would solve this, otherwise we've got to have the extra setup in the config file, or tons of validation in the ExtensionService. 
2b. What is the ExtensionService buying us here? I think the original intent was to allow a place to add status conditions, but I'm still unclear as to how that happens.

Reply:
> 1. This assumes a single cluster will only have one RLS? (Maybe good to call out in the goals). I'm thinking possibly one team may want their own implementation of a global RLS different than another.

The challenge is that (AFAICT) you can only have one RLS per HTTP connection manager, and all non-TLS virtual hosts share the same HTTP connection manager. External auth ran into the same constraint, and opted to support external auth *only* for TLS vhosts, so that you could have a separate auth service per vhost. We *could* make that same choice for rate limiting - allow global rate limiting only for TLS vhosts - but that seems a little strange to me. Definitely worth some discussion.

> 2. It's unfortunate to need another config place. It seems like the ExtensionService should be where you define the domain, denyOnFailure, etc. Having an `RateLimitExtensionService` and a `AuthorizationExtensionService` would solve this, otherwise we've got to have the extra setup in the config file, or tons of validation in the ExtensionService.

Yeah - *if* we could configure an RLS per vhost, then these things would go into a struct as part of HTTPProxy, like AuthorizationServer today. See previous comment, though.

>    2b. What is the ExtensionService buying us here? I think the original intent was to allow a place to add status conditions, but I'm still unclear as to how that happens.

Basically gets the programming of the Envoy cluster for the RLS for free. Other than that, there's not currently any infrastructure for doing health checks, etc.

Reply:
> The challenge is that (AFAICT) you can only have one RLS per HTTP connection manager
That makes sense, should just call it out in the doc so that is clear to readers. I don't think we should add a requirement that only TLS'd vhosts will get this service, for auth it makes sense to keep the security boundaries tight, but for rate limiting I don't think it's a big deal other than possibly someone could fake the requests to get around rate limiting? 

> Yeah - if we could configure an RLS per vhost, then these things would go into a struct as part of HTTPProxy, like AuthorizationServer today. See previous comment, though.

You could move some of the configuration items to the vhost, like domain, etc. Also might be a place to have global ratelimiting configuration per vhost.

Reply:
I'm increasingly thinking that we should bring some per-ExtensionService properties into the ExtensionService,  like we already do with timeouts and load balancing policy. However, each type of ExtensionService almost always has extra parameters that belong to the service (in this case, they're currently listed as in the configuration file, and in the tracing case, they would probably need to be part of the vhost configuration currently), and each type of thing has a different set of things that need to be configured.

In this case,  it's these two fields here: `domain` and `failOpen`, and when we do tracing (ie #399 ) there will be other service-specific things, that will be the same for every reference of that service. If we moved to doing this, the config file would just be for the ExtensionServiceRef rather than for config details, which would allow those config details to be flipped without Contour redeploys. Here I'm thinking of flipping between `failOpen` states - that seems like a change that you would want to be able to back out quickly in the event of a problem.

If we could add some sort of (hah) extension point into ExtensionService, that would allow you to supply a set of fields or something, we could move this config into the relevant ExtensionService (where I feel like it kind of belongs, since it's really a property of that service).

It does seem like the second time we use ExtensionService is a good time to think about its general properties. However, I'd be okay with putting this as an open question, to be determined during implementation because it requires some experimentation.

Reply:
> If we could add some sort of (hah) extension point into ExtensionService, that would allow you to supply a set of fields or something, we could move this config into the relevant ExtensionService (where I feel like it kind of belongs, since it's really a property of that service).

This was my original suggestion when first building out ExensionService, don't make it generic, but make it specific to the need you are configuring. So you'd have an `ExtensionServiceAuth` and `ExtensionServiceRateLimit`, `ExtensionServiceOauth2`, etc, etc. This way you can be clear about fields that need configured and not rely on conditions to say, "actually user, that field isn't meant for this context so don't set it". 

There's a trade-off in complexity here. The overhead of having more CRDs vs a nicer user experience.